File: python/notary_public/receipt.py
"""
Local signed-license-receipt storage. Two locations, searched in order:
  1. $INTERCHAINED_LICENSE_FILE  (explicit override)
  2. <cwd>/.nedb/license.json     (project-local)
  3. ~/.interchained/licenses/<product>.json  (user/global)
"""
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Optional


def user_receipt_path(product: str) -> Path:
    return Path.home() / ".interchained" / "licenses" / f"{product}.json"


def project_receipt_path(cwd: Optional[str] = None) -> Path:
    base = Path(cwd) if cwd else Path.cwd()
    return base / ".nedb" / "license.json"


def find_receipt(product: str, opts: dict | None = None) -> Optional[dict]:
    """
    Search receipt locations in priority order.
    Returns {"receipt": <dict>, "path": <str>} or None.
    """
    opts = opts or {}
    candidates = []
    env_path = os.environ.get("INTERCHAINED_LICENSE_FILE")
    if env_path:
        candidates.append(env_path)
    candidates.append(str(project_receipt_path(opts.get("cwd"))))
    candidates.append(str(user_receipt_path(product)))

    for p in candidates:
        if not p:
            continue
        path = Path(p)
        try:
            if path.exists():
                r = json.loads(path.read_text("utf-8"))
                if not product or r.get("product") == product:
                    return {"receipt": r, "path": str(path)}
        except Exception:
            pass  # malformed receipt — keep searching
    return None


def write_user_receipt(product: str, receipt: dict) -> str:
    p = user_receipt_path(product)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(receipt, indent=2), encoding="utf-8")
    return str(p)

File: src/receipt.js
"use strict";
// Local signed-license-receipt storage. Two locations, searched in order:
//   1. $INTERCHAINED_LICENSE_FILE  (explicit override)
//   2. <cwd>/.nedb/license.json     (project-local)
//   3. ~/.interchained/licenses/<product>.json  (user/global)
const fs = require("fs");
const os = require("os");
const path = require("path");

function userReceiptPath(product) {
  return path.join(os.homedir(), ".interchained", "licenses", `${product}.json`);
}
function projectReceiptPath(cwd) {
  return path.join(cwd || process.cwd(), ".nedb", "license.json");
}

function findReceipt(product, opts = {}) {
  const candidates = [];
  if (process.env.INTERCHAINED_LICENSE_FILE) candidates.push(process.env.INTERCHAINED_LICENSE_FILE);
  candidates.push(projectReceiptPath(opts.cwd));
  candidates.push(userReceiptPath(product));
  for (const p of candidates) {
    try {
      if (p && fs.existsSync(p)) {
        const r = JSON.parse(fs.readFileSync(p, "utf8"));
        if (!product || r.product === product) return { receipt: r, path: p };
      }
    } catch {
      /* malformed receipt — keep searching */
    }
  }
  return null;
}

function writeUserReceipt(product, receipt) {
  const p = userReceiptPath(product);
  fs.mkdirSync(path.dirname(p), { recursive: true });
  fs.writeFileSync(p, JSON.stringify(receipt, null, 2));
  return p;
}

module.exports = { userReceiptPath, projectReceiptPath, findReceipt, writeUserReceipt };
